File: src-tauri/mangadex/src/subscription/download_state/chapter.rs
use crate::{
    subscription::utils::WatchSubscriptionStream,
    utils::{
        download_state_rx::{get_download_state_rx, NextTaskValue},
        traits_utils::{MangadexAsyncGraphQLContextExt, MangadexTauriManagerExt},
    },
    Result,
};
use actix::Addr;
use async_graphql::{Context, Object, SimpleObject, Subscription};
use async_stream::stream;
use eureka_mmanager::{
    download::{
        chapter::{
            task::{ChapterDownloadTaskState, ChapterDownloadingState as DownloadingState},
            ChapterDownloadManager,
        },
        GetManager,
    },
    prelude::{ChapterDownloadTask, TaskManagerAddr},
    DownloadManager, OwnedError,
};
use tauri::{Manager, Runtime};
use tokio::{select, sync::watch::Receiver};
use tokio_stream::Stream;
use uuid::Uuid;

#[derive(Debug, Clone)]
pub enum ChapterDownloadState {
    Pending,
    Done,
    Downloading(DownloadingState),
    Canceled,
    Error(OwnedError),
    OfflineAppStateNotLoaded,
}

impl NextTaskValue for ChapterDownloadState {
    type DownloadingState = DownloadingState;

    fn pending() -> Self {
        Self::Pending
    }

    fn downloading(value: Self::DownloadingState) -> Self {
        Self::Downloading(value)
    }

    fn error(error: OwnedError) -> Self {
        Self::Error(error)
    }

    fn done() -> Self {
        Self::Done
    }

    fn canceled() -> Self {
        Self::Canceled
    }

    fn offline_app_state_not_loaded() -> Self {
        Self::OfflineAppStateNotLoaded
    }

    fn is_pending(&self) -> bool {
        matches!(self, Self::Pending)
    }

    fn is_done(&self) -> bool {
        matches!(self, Self::Done)
    }

    fn is_canceled(&self) -> bool {
        matches!(self, Self::Canceled)
    }

    fn is_offline_app_state_not_loaded(&self) -> bool {
        matches!(self, Self::OfflineAppStateNotLoaded)
    }
}

impl From<ChapterDownloadTaskState> for ChapterDownloadState {
    fn from(value: ChapterDownloadTaskState) -> Self {
        match value {
            eureka_mmanager::download::state::DownloadTaskState::Pending => Self::Pending,
            eureka_mmanager::download::state::DownloadTaskState::Loading(s) => Self::Downloading(s),
            eureka_mmanager::download::state::DownloadTaskState::Error(owned_error) => {
                Self::Error(owned_error)
            }
            eureka_mmanager::download::state::DownloadTaskState::Done(_) => Self::Done,
            eureka_mmanager::download::state::DownloadTaskState::Canceled => Self::Canceled,
        }
    }
}

pub struct ChapterDownloadingState(DownloadingState);

impl From<DownloadingState> for ChapterDownloadingState {
    fn from(value: DownloadingState) -> Self {
        Self(value)
    }
}

#[derive(SimpleObject)]
pub struct ChapterImageFetchingStatus {
    filename: String,
    index: usize,
    len: usize,
}

#[Object]
impl ChapterDownloadingState {
    pub async fn is_preloading(&self) -> bool {
        matches!(self.0, DownloadingState::Preloading)
    }
    pub async fn is_fetching_data(&self) -> bool {
        matches!(self.0, DownloadingState::FetchingData)
    }
    pub async fn fetching_image(&self) -> Option<ChapterImageFetchingStatus> {
        if let DownloadingState::FetchingImage {
            filename,
            index,
            len,
        } = self.0.clone()
        {
            Some(ChapterImageFetchingStatus {
                filename,
                index,
                len,
            })
        } else {
            None
        }
    }
    pub async fn is_fetching_at_home_data(&self) -> bool {
        matches!(self.0, DownloadingState::FetchingAtHomeData)
    }
}

#[Object]
impl ChapterDownloadState {
    pub async fn is_pending(&self) -> bool {
        matches!(self, Self::Pending)
    }
    pub async fn is_done(&self) -> bool {
        matches!(self, Self::Done)
    }
    pub async fn is_canceled(&self) -> bool {
        matches!(self, Self::Canceled)
    }
    pub async fn is_offline_app_state_not_loaded(&self) -> bool {
        matches!(self, Self::OfflineAppStateNotLoaded)
    }
    pub async fn error(&self) -> Option<String> {
        if let Self::Error(err) = self {
            Some(err.to_string())
        } else {
            None
        }
    }
    pub async fn downloading(&self) -> Option<ChapterDownloadingState> {
        if let Self::Downloading(d) = self {
            Some(d.clone().into())
        } else {
            None
        }
    }
}

fn get_chapter_download_state_rx<R: Runtime, M: Manager<R> + Clone + Send + 'static>(
    app: &M,
    id: Uuid,
) -> crate::Result<Receiver<ChapterDownloadState>> {
    get_download_state_rx::<ChapterDownloadManager, ChapterDownloadTask, _, R, M>(app, id)
}

pub struct ChapterDownloadSubs;

#[Subscription]
impl ChapterDownloadSubs {
    pub async fn listen_to_chapter_tasks<'ctx>(
        &'ctx self,
        ctx: &'ctx Context<'ctx>,
    ) -> Result<impl Stream<Item = Vec<Uuid>> + 'ctx> {
        let window = ctx.get_window::<tauri::Wry>()?.clone();
        let maybe_offline = (*window.get_offline_app_state()?).clone();
        let offline_read = maybe_offline
            .read()
            .await
            .as_ref()
            .map(|e| e.app_state.clone())
            .ok_or(crate::Error::OfflineAppStateNotLoaded)?;
        let manager =
            <Addr<DownloadManager> as GetManager<ChapterDownloadManager>>::get(&offline_read)
                .await?;
        let notify = manager.notify().await?;
        let stream = stream! {
            if let Ok(tasks) = manager.tasks_id().await {
                yield tasks
            }
            loop {
                select! {
                    _ = notify.notified() => {
                        if let Ok(tasks) = manager.tasks_id().await {
                            yield tasks
                        }
                    }
                    else => break
                }
            }
        };
        Ok(stream)
    }
    pub async fn listen_to_download_state<'ctx>(
        &'ctx self,
        ctx: &'ctx Context<'ctx>,
        chapter_id: Uuid,
    ) -> Result<impl Stream<Item = ChapterDownloadState> + 'ctx> {
        let window = ctx.get_window::<tauri::Wry>()?.clone();
        Ok(WatchSubscriptionStream::new(get_chapter_download_state_rx(
            &window, chapter_id,
        )?))
        /* let stream = stream! {
            let mut is_readed = false;
            loop {
                select! {
                    Some(mounted) = is_mounted.next() => {
                        if mounted {
                            let olar = maybe_offline.read().await;
                            if let Some(offline) = (*olar).as_ref() {
                                let _ = tx.send_replace(Some(offline.app_state.clone()));
                            }else {
                                yield ChapterDownloadState::OfflineAppStateNotLoaded;
                            }
                        }else {
                            let _ = tx.send_replace(None);
                            is_readed = false;
                            yield ChapterDownloadState::OfflineAppStateNotLoaded;
                        }
                    }
                    Some(manager) = async{ (*rx.borrow()).clone() } => {
                        match GetManager::<ChapterDownloadManager>::get(&manager).await {
                            Ok(manager) => {
                                match manager.new_task(ChapterDownloadMessage::new(chapter_id)).await {
                                    Ok(task) => {
                                        match task.subscribe().await {
                                            Ok(mut sub) => {
                                                if is_readed {
                                                    let _ = sub.changed().await;
                                                }else {
                                                    is_readed = true;
                                                }
                                                let data: ChapterDownloadState = {
                                                    (*sub.borrow()).clone().into()
                                                };
                                                yield data
                                            },
                                            Err(err) => yield ChapterDownloadState::Error(err.into())
                                        }
                                    },
                                    Err(err) => {
                                        yield ChapterDownloadState::Error(ManagerError::MailBox(err).into())
                                    }
                                }
                            },
                            Err(err) => {
                                yield ChapterDownloadState::Error(ManagerError::MailBox(err).into())
                            }
                        }
                    }
                    else => break
                }
            }
            println!("Leaves streams")
        };
        Ok(stream)
         */
    }
}

File: apps/mangadex/lib/download/chapter.ts
import { graphql } from "@mangadex/gql";
import type { DownloadMode } from "@mangadex/gql/graphql";
import { client } from "@mangadex/gql/urql";
import { subscriptionStore } from "@urql/svelte";
import { derived, type Readable } from "svelte/store";

const download_mutation = graphql(`
	mutation downloadChapterMutation($id: UUID!, $quality: DownloadMode) {
		chapter {
			download(id: $id, quality: $quality) {
				hasFailed
				isDownloaded
			}
		}
	}
`);

const canceled_download_mutation = graphql(`
	mutation cancelDownloadChapterMutation($id: UUID!) {
		chapter {
			cancelDownload(id: $id)
		}
	}
`);

const subscription = graphql(`
	subscription chapterDownloadState($id: UUID!) {
		watchChapterDownloadState(chapterId: $id) {
			isPending
			isDone
			isCanceled
			isOfflineAppStateNotLoaded
			error
			downloading {
				isPreloading
				isFetchingData
				fetchingImage {
					filename
					index
					len
				}
				isFetchingAtHomeData
			}
		}
	}
`)

const remove_chapter_mutation = graphql(`
	mutation removeDownloadedChapter($id: UUID!) {
		chapter {
			remove(id: $id)
		}
	}
`);

export enum ChapterDownloadState {
	Pending,
	Done,
	Canceled,
	OfflineAppStateNotLoaded,
	Error,
	Preloading,
	FetchingAtHomeData,
	FetchingImages,
	FetchingData
}

export class ChapterDownload {
	private chapterId: string;
	private mode?: DownloadMode;
	/**
	 *
	 */
	constructor(chapterId: string, mode?: DownloadMode) {
		this.chapterId = chapterId,
			this.mode = mode;
	}
	public set quality(val: DownloadMode) {
		this.mode = val
	}
	public get quality(): DownloadMode | undefined {
		return this.mode;
	}
	public sub_raw_state() {
		return subscriptionStore({
			client,
			query: subscription,
			variables: {
				id: this.chapterId
			}
		})
	}
	public state(): Readable<ChapterDownloadState> {
		return derived(this.sub_raw_state(), (result) => {
			if (result.data) {
				const data = result.data.watchChapterDownloadState;
				if (data.downloading) {
					const downloading = data.downloading
					if (downloading.fetchingImage) {
						return ChapterDownloadState.FetchingImages
					} else if (downloading.isFetchingAtHomeData) {
						return ChapterDownloadState.FetchingAtHomeData
					} else if (downloading.isFetchingData) {
						return ChapterDownloadState.FetchingData
					} else {
						return ChapterDownloadState.Preloading
					}
				} else if (data.error) {
					return ChapterDownloadState.Error
				} else if (data.isCanceled) {
					return ChapterDownloadState.Canceled
				} else if (data.isDone) {
					return ChapterDownloadState.Done
				} else if (data.isOfflineAppStateNotLoaded) {
					return ChapterDownloadState.OfflineAppStateNotLoaded
				} else {
					return ChapterDownloadState.Pending
				}
			} else if (result.error) {
				return ChapterDownloadState.Error;
			} else {
				return ChapterDownloadState.Pending;
			}
		})
	}
	public is_downloading() {
		return derived(this.sub_raw_state(), (result) => {
			const downloading = result.data?.watchChapterDownloadState.downloading;
			return downloading != undefined && downloading != null
		})
	}
	public static download_mutation() {
		return download_mutation
	}
	public static subscription() {
		return subscription
	}
	public static cancel_dowload_mutation() {
		return canceled_download_mutation;
	}
	public static remove_chapter_mutation() {
		return remove_chapter_mutation
	}
	public images_state() {
		return derived(this.sub_raw_state(), (result) => {
			result.data?.watchChapterDownloadState.downloading?.fetchingImage
		})
	}
	public error() {
		return derived(this.sub_raw_state(), (result) => {
			if (result.error) {
				return result.error
			} else if (result.data?.watchChapterDownloadState.error) {
				return new Error(result.data.watchChapterDownloadState.error)
			}
		})
	}
	public async download() {
		return await client.mutation(ChapterDownload.download_mutation(), {
			id: this.chapterId,
			quality: this.mode
		}).toPromise();
	}
	public async cancel() {
		return await client.mutation(ChapterDownload.cancel_dowload_mutation(), {
			id: this.chapterId
		}).toPromise();
	}
	public async remove() {
		return await client.mutation(ChapterDownload.remove_chapter_mutation(), {
			id: this.chapterId
		}).toPromise();
	}
	public has_failed() {
		return derived(this.state(), (result) => {
			switch (result) {
				case ChapterDownloadState.Error | ChapterDownloadState.Canceled:
					return true;

				default:
					return false;
			}
		})
	}
	public is_downloaded() {
		return derived(this.state(), (result) => {
			switch (result) {
				case ChapterDownloadState.Done:
					return true;

				default:
					return false;
			}
		})
	}
}